### PR TITLE
glibc: depends on gawk for Linuxbrew

### PR DESCRIPTION
**Pull request by DoomHammer**
- [X] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/homebrew-core/blob/master/.github/CONTRIBUTING.md) document?
- [X] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [X] Have you built your formula locally prior to submission with `brew install <formula>` (where `<formula>` is the name of the formula you're submitting)?
- [X] Does your submission pass `brew audit --strict --online <formula>` (after doing `brew install <formula>`)?

-----

Depends on https://github.com/Linuxbrew/brew/pull/34 .

### CODE REVIEW

**Comment by sjackman** [comment]
For the dependencies of `glibc`, see http://www.gnu.org/software/libc/manual/html_node/Tools-for-Compilation.html
```
GNU make 3.79 or newer
GCC 4.7 or newer
GNU binutils 2.22 or later
GNU texinfo 4.7 or later
GNU awk 3.1.2, or higher
GNU sed 3.02 or newer
```

File: Formula/glibc.rb [hunk 1]
```diff
@@ -1,3 +1,20 @@
...
+  fatal true
+  default_formula "gawk"
+
+  satisfy :build_env => false do
+    begin
+      system "gawk", "--version"
+    rescue
+      false
+    end
+  end
```
Comment (sjackman) [inline Formula/glibc.rb]: The usual approach is simply:
```ruby
  satisfy :build_env => false { which("gawk") }
```
See https://github.com/Linuxbrew/brew/blob/master/Library/Homebrew/requirements.rb#L125-L130

File: Formula/glibc.rb [hunk 2]
```diff
@@ -1,3 +1,20 @@
...
+
+  def message
+    "Gawk is required to install this formula."
+  end
```
Comment (sjackman) [inline Formula/glibc.rb]: `def message` is not needed. The default message should be fine.

**Comment by DoomHammer** [comment]
Fixed

**Comment by sjackman** [comment]
:+1: Merge it! Remember to delete the `revision` before merging.

**Comment by sjackman** [comment]
Interesting error from Travis (unrelated to your patch):
```
/usr/bin/gcc-4.6 -B/home/linuxbrew/.linuxbrew/Cellar/binutils/2.26/bin/ ../sysdeps/x86_64/fpu/multiarch/e_atan2.c -c -std=gnu99 -fgnu89-inline -fno-stack-protector -Os -Wall -Winline -Wwrite-strings -fmerge-all-constants -frounding-math -march=native -pipe -w -Wstrict-prototypes      -Wno-uninitialized -U_FORTIFY_SOURCE -isystem/home/linuxbrew/.linuxbrew/include -D__NO_MATH_INLINES -D__LIBC_INTERNAL_MATH_INLINES -I../include -I/tmp/glibc-20160510-25814-160ebru/glibc-2.19/build/math  -I/tmp/glibc-20160510-25814-160ebru/glibc-2.19/build  -I../sysdeps/unix/sysv/linux/x86_64/64/nptl  -I../sysdeps/unix/sysv/linux/x86_64/64  -I../nptl/sysdeps/unix/sysv/linux/x86_64  -I../nptl/sysdeps/unix/sysv/linux/x86  -I../sysdeps/unix/sysv/linux/x86  -I../sysdeps/unix/sysv/linux/x86_64  -I../sysdeps/unix/sysv/linux/wordsize-64  -I../nptl/sysdeps/unix/sysv/linux  -I../nptl/sysdeps/pthread  -I../sysdeps/pthread  -I../ports/sysdeps/unix/sysv/linux  -I../sysdeps/unix/sysv/linux  -I../sysdeps/gnu  -I../sysdeps/unix/inet  -I../nptl/sysdeps/unix/sysv  -I../ports/sysdeps/unix/sysv  -I../sysdeps/unix/sysv  -I../sysdeps/unix/x86_64  -I../nptl/sysdeps/unix  -I../ports/sysdeps/unix  -I../sysdeps/unix  -I../sysdeps/posix  -I../nptl/sysdeps/x86_64/64  -I../sysdeps/x86_64/64  -I../sysdeps/x86_64/fpu/multiarch  -I../sysdeps/x86_64/fpu  -I../sysdeps/x86/fpu  -I../sysdeps/x86_64/multiarch  -I../nptl/sysdeps/x86_64  -I../sysdeps/x86_64  -I../sysdeps/x86  -I../sysdeps/ieee754/ldbl-96  -I../sysdeps/ieee754/dbl-64/wordsize-64  -I../sysdeps/ieee754/dbl-64  -I../sysdeps/ieee754/flt-32  -I../sysdeps/wordsize-64  -I../sysdeps/ieee754  -I../sysdeps/generic  -I../nptl  -I../ports  -I.. -I../libio -I. -nostdinc -isystem /usr/lib/gcc/x86_64-linux-gnu/4.6/include -isystem /usr/lib/gcc/x86_64-linux-gnu/4.6/include-fixed -isystem /home/linuxbrew/.linuxbrew/Cellar/linux-headers/3.18.27/include  -D_LIBC_REENTRANT -include ../include/libc-symbols.h   -DNOT_IN_libc=1 -DIS_IN_libm=1 -DIN_LIB=libm    -o /tmp/glibc-20160510-25814-160ebru/glibc-2.19/build/math/e_atan2.o -MD -MP -MF /tmp/glibc-20160510-25814-160ebru/glibc-2.19/build/math/e_atan2.o.dt -MT /tmp/glibc-20160510-25814-160ebru/glibc-2.19/build/math/e_atan2.o
{standard input}: Assembler messages:
{standard input}:42: Error: no such instruction: `vfmadd312sd .LC13(%rip),%xmm0,%xmm1'
```
https://travis-ci.org/Linuxbrew/homebrew-core/builds/129089906#L774

**Comment by sjackman** [comment]
I spoke too soon. There's a couple problems:

```
==> Installing glibc dependency: gawk
```
https://travis-ci.org/Linuxbrew/homebrew-core/builds/129089906#L774

Travis host has `awk` installed. Why is it installing Linuxbrew `gawk? Does it have a different `awk` than `gawk`, maybe `mawk`?

**Comment by sjackman** [comment]
```
==> Verifying gawk-4.1.3.x86_64_linux.bottle.tar.gz checksum
==> Pouring gawk-4.1.3.x86_64_linux.bottle.tar.gz
tar xzf /home/linuxbrew/.cache/Homebrew/gawk-4.1.3.x86_64_linux.bottle.tar.gz
/home/linuxbrew/.linuxbrew/Cellar/patchelf/0.9/bin/patchelf --set-rpath /home/linuxbrew/.linuxbrew/lib --set-interpreter /home/linuxbrew/.linuxbrew/lib/ld.so /home/linuxbrew/.linuxbrew/Cellar/gawk/4.1.3/bin/gawk 
/home/linuxbrew/.linuxbrew/Cellar/patchelf/0.9/bin/patchelf --set-rpath /home/linuxbrew/.linuxbrew/lib --set-interpreter /home/linuxbrew/.linuxbrew/lib/ld.so /home/linuxbrew/.linuxbrew/Cellar/gawk/4.1.3/bin/gawk-4.1.3 
cannot find section .interp
Error: Failure while executing: /home/linuxbrew/.linuxbrew/Cellar/patchelf/0.9/bin/patchelf --set-rpath /home/linuxbrew/.linuxbrew/lib 
Warning: Bottle installation failed: building from source.
```

The `gawk` bottle fails to pour when running `patchelf`. Is it a static executable or otherwise weird?
https://travis-ci.org/Linuxbrew/homebrew-core/builds/129089906#L753

**Comment by sjackman** [comment]
Fun cyclic dependency here:
```
==> Installing glibc dependency: gawk
…
Warning: Bottle installation failed: building from source.
==> Installing dependencies for gawk: glibc, xz, gmp, gpatch, mpfr, libmpc, isl, gcc
==> Installing gawk dependency: glibc
```
https://travis-ci.org/Linuxbrew/homebrew-core/builds/129089906#L756

**Comment by DoomHammer** [comment]
Debian usually has `mawk` installed, though `glibc` says it needs `gawk` at least 3.sth. WHat should we do about that?

**Comment by sjackman** [comment]
It's been building just fine so far, so I'm guessing that `mawk` is good enough. Let's change the test to `which("awk")`.

**Comment by DoomHammer** [comment]
The whole idea of this requirement, IIRC, was that Debian Lenny had `awk` installed but it was insufficient to build `glibc`. Back to the drawing board, I think?

**Comment by sjackman** [comment]
Interesting. So Ubuntu 12's `awk` is good enough to build `glibc`, but Debian Lenny's `awk` is not good enough? Can you report `awk --version` for both distros?

**Comment by DoomHammer** [comment]
```
awk --version
awk: not an option: --version
```

```
dpkg -l|grep awk
ii  mawk                          1.3.3-11.1             a pattern scanning and text processing langu
```

for lenny

**Comment by DoomHammer** [comment]
`1.3.3.17` for Ubuntu 12.04 and 12.10.

**Comment by sjackman** [comment]
Travis CI running Ubuntu 12 has been building `glibc` just fine. I wonder which `awk` it has been using.

**Comment by sjackman** [comment]
Recently Travis has been using the `glibc` bottle, but it used to build it from source.

**Comment by DoomHammer** [comment]
My memory is not very reliable, so I do not recall what exactly it was that lead us to do this. Thus, I will try to recreate the problematic image so anyone can play with it. Either this or we'll bury the PR ;)

**Comment by sjackman** [comment]
This PR is useful I think if the system has no `awk` installed whatsoever.

File: Formula/glibc.rb [hunk 3]
```diff
@@ -1,9 +1,27 @@
...
+end
+
 class Glibc < Formula
   desc "The GNU C Library"
   homepage "https://www.gnu.org/software/libc/download.html"
   url "http://ftpmirror.gnu.org/glibc/glibc-2.19.tar.bz2"
   sha256 "2e293f714187044633264cd9ce0183c70c3aa960a2f77812a6390a3822694d15"
   # tag "linuxbrew"
+  revision 1 # remove this before merging`
```
Comment (sjackman) [inline Formula/glibc.rb]: Change this to
```ruby
revision 1 if OS.linux?
```

Comment (DoomHammer) [inline Formula/glibc.rb]: Isn't it always `OS.linux?`?

Comment (sjackman) [inline Formula/glibc.rb]: Yes, sorry. I must have been tired.

**Comment by sjackman** [comment]
I've learned fairly recently that Requirements don't work as I thought they did. When a formula is installed from a bottle, its requirements become hard dependencies, which is not what we want here.

**Comment by DoomHammer** [comment]
Back to the drawing board, then?

**Comment by sjackman** [comment]
Yep. I'd like to have a Requirements class that works the way I thought it did, without any special behaviour when installing from a bottle.

**Comment by DoomHammer** [comment]
So what shall we do with this PR? Close it for the time being?

**Comment by sjackman** [comment]
Yes, let's close it for now (without prejudice). Although `gawk` depends on `gmp mpfr ncurses readline` and so is not a good candidate for a `glibc` dependency, `mawk` has no dependencies and could be used when the system does not provide `awk` to compile `glibc`.

**Pull request closed without merging**